File: internal/models/contact_campaign_state.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// ContactCampaignState is one campaign a contact is a lead of: steps with
// this contact's progress, the derived lead status, and the next action.
type ContactCampaignState struct {
	CampaignID     uuid.UUID `json:"campaign_id"`
	CampaignName   string    `json:"campaign_name"`
	CampaignStatus string    `json:"campaign_status"`
	// LeadStatus follows ContactCampaignProgress.Status exactly.
	LeadStatus string `json:"lead_status"`

	// SenderID / SenderEmail are the mailbox this lead's whole sequence sends
	// from, fixed when its first email went out. Empty until then: rotation
	// picks it for the first step and every follow-up follows.
	SenderID    *uuid.UUID `json:"sender_id,omitempty"`
	SenderEmail string     `json:"sender_email,omitempty"`
	// FailureReason is the worker's reason for the last failed send.
	FailureReason string `json:"failure_reason,omitempty"`

	Steps          []ContactCampaignStep `json:"steps"`
	CompletedSteps int                   `json:"completed_steps"`
	TotalSteps     int                   `json:"total_steps"`

	// CurrentStep is the latest step sent.
	CurrentStep  *ContactCampaignStep `json:"current_step,omitempty"`
	LastAction   string               `json:"last_action,omitempty"`
	LastActionAt *time.Time           `json:"last_action_at,omitempty"`

	// Next is nil once the flow has ended for the contact; EndedReason says why.
	Next        *ContactNextAction `json:"next,omitempty"`
	EndedReason string             `json:"ended_reason,omitempty"`
}

// ContactCampaignStep is one flow node with this contact's progress on it.
type ContactCampaignStep struct {
	ID       uuid.UUID `json:"id"`
	Label    string    `json:"label"`
	Kind     string    `json:"kind"`
	Position int       `json:"position"`
	Subject  string    `json:"subject,omitempty"`

	SentAt    *time.Time `json:"sent_at,omitempty"`
	OpenedAt  *time.Time `json:"opened_at,omitempty"`
	ClickedAt *time.Time `json:"clicked_at,omitempty"`
	RepliedAt *time.Time `json:"replied_at,omitempty"`
	BouncedAt *time.Time `json:"bounced_at,omitempty"`
	FailedAt  *time.Time `json:"failed_at,omitempty"`
	// Attempts counts failed sends; InFlight is a reservation with no result yet.
	Attempts int  `json:"attempts,omitempty"`
	InFlight bool `json:"in_flight,omitempty"`
}

// ContactNextActionState says how firm the next action's timing is; only
// "due" carries the campaign's next wakeup.
type ContactNextActionState string

const (
	NextActionDue     ContactNextActionState = "due"
	NextActionWaiting ContactNextActionState = "waiting"
	NextActionPaused  ContactNextActionState = "paused"
	NextActionBlocked ContactNextActionState = "blocked"
)

// ContactNextAction is derived on read through the scheduler; never stored.
type ContactNextAction struct {
	// StepID is nil while a branch condition is still undecided.
	StepID    *uuid.UUID `json:"step_id,omitempty"`
	StepLabel string     `json:"step_label"`
	Kind      string     `json:"kind,omitempty"`
	Subject   string     `json:"subject,omitempty"`

	State ContactNextActionState `json:"state"`
	// ScheduledAt is set only when due, and is then when the campaign's chain
	// next wakes up and works its queue — a stored wakeup, so it reads the same
	// on every refresh instead of a freshly paced slot that walked forward
	// every time the drawer polled (issue #437). NotBefore is the earliest the
	// hard constraints allow; Constraint names the gate in user-facing words.
	ScheduledAt *time.Time `json:"scheduled_at,omitempty"`
	NotBefore   *time.Time `json:"not_before,omitempty"`
	Constraint  string     `json:"constraint,omitempty"`
}

type ContactCampaignStatesResult struct {
	Data []ContactCampaignState `json:"data"`
}

File: web/src/lib/api/models/app/contacts/ContactCampaignState.ts
import type { LeadStatus } from "./Contact";

// One campaign a contact belongs to, as the Activity tab's campaign panel
// shows it: the flow with this contact's progress, the derived lead status,
// and a scheduler-backed next action.
export interface ContactCampaignStep {
    id: string;
    label: string;
    kind: string;
    position: number;
    subject?: string;

    sent_at?: string | null;
    opened_at?: string | null;
    clicked_at?: string | null;
    replied_at?: string | null;
    bounced_at?: string | null;
    failed_at?: string | null;
    attempts?: number;
    in_flight?: boolean;
}

// due: the step is due; scheduled_at is when the campaign next works its queue.
// waiting: a hard constraint holds it back; not_before is the earliest.
// paused: the campaign is not active.
// blocked: the campaign cannot send at all right now.
export type ContactNextActionState = "due" | "waiting" | "paused" | "blocked";

export interface ContactNextAction {
    // Absent while a condition window is open: the next step depends on how
    // the contact responds.
    step_id?: string | null;
    step_label: string;
    kind?: string;
    subject?: string;

    state: ContactNextActionState;
    // Only when due now, and then it is the campaign chain's own next wakeup,
    // not a slot reserved for this contact: leads queued ahead can still push
    // this step to a later pass. Absent while the chain is being re-seeded.
    scheduled_at?: string | null;
    not_before?: string | null;
    constraint?: string;
}

export default interface ContactCampaignState {
    campaign_id: string;
    campaign_name: string;
    campaign_status: string;
    lead_status: LeadStatus;
    failure_reason?: string;

    // The mailbox this lead's whole sequence sends from. Rotation picks it for
    // the first email and every follow-up keeps it, so the contact always hears
    // from one address. Absent until the first email goes out.
    sender_id?: string | null;
    sender_email?: string;

    steps: ContactCampaignStep[];
    completed_steps: number;
    total_steps: number;

    current_step?: ContactCampaignStep | null;
    last_action?: string;
    last_action_at?: string | null;

    next?: ContactNextAction | null;
    ended_reason?: string;
}

export interface ContactCampaignStatesResult {
    data: ContactCampaignState[];
}
